### PR TITLE
UCM/TEST: Use hash instead of array to store mmap'ed pointers.

### LINKED ISSUE #1974: UCM consumes too much time traversing mmaped_pointers array
When the application allocates many blocks with mmap (because the allocation size is larger than MALLOC_MMAP_THRESHOLD), the array search is linear and takes a long time. Every time such block is released, need to check if it was allocated by UCM or not, and the linear search takes a long time.

### PR DESCRIPTION
**Pull request by yosefe**
Fixes #1974

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5058/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5061/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5063/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich @evgeny-leksikov pls take a look

File: src/ucm/util/sys.h [hunk 1]
```diff
@@ -11,6 +11,17 @@
 #include <stddef.h>
 
 
+/*
+ * Substitutes for glibc memory allocation routings, which take memory
```
Comment (brminich) [inline src/ucm/util/sys.h]: routins

File: test/gtest/ucm/malloc_hook.cc [hunk 2]
```diff
@@ -355,28 +356,61 @@ UCS_TEST_F(malloc_hook, fork) {
...
+            break;
+        }
+    }
+
+    double measure_alloc_time(size_t size, unsigned iters)
+    {
+        ucs_time_t start_time = ucs_get_time();
+        for (unsigned i = 0; i < iters; ++i) {
+            void *ptr = malloc(size);
+            *(char*)ptr = '5';
```
Comment (brminich) [inline test/gtest/ucm/malloc_hook.cc]: you just touch the memory and 5 has no any special meaning, right?

File: test/gtest/ucm/malloc_hook.cc [hunk 3]
```diff
@@ -462,6 +492,63 @@ UCS_TEST_F(malloc_hook_cplusplus, mallopt) {
...
+    EXPECT_GE(m_mapped_size, large_blocks * size);
+
+    /* Allocate many large strings to trigger mmap() based allocation. */
+    for (unsigned i = 0; i < count; ++i) {
+        std::string str(size, 't');
+        ptrs.push_back(str);
+        ++large_blocks;
+    }
+
+    /* Measure allocation time with many large blocks on th heap */
```
Comment (brminich) [inline test/gtest/ucm/malloc_hook.cc]: the

File: test/gtest/ucm/malloc_hook.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/ucm/malloc_hook.cc]: yes, i want to prevent the compiler from removing that code

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5068/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5074/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5075/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5076/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich fixed

**Comment by brminich** [comment]
:+1:

File: test/gtest/ucm/malloc_hook.cc (hunk 2 above)
Comment (shamisp) [reply test/gtest/ucm/malloc_hook.cc]: add comment, otherwise somebody else will remove it

**Comment by yosefe** [comment]
@shamisp fixed+squashed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5084/ for details (Mellanox internal link).

**Pull request merged**